rpcclient/roundrobin.go

```diff
@@ -0,0 +1,434 @@
+	// client is marked as Shutdown.
+	ErrNoUsableConnAvailable = errors.New("no Connection is available for use, all are marked as Shutdown")
+)
+
+// ConnectionState indicates the current state of connection.
+type ConnectionState int
+
+// String would return a string representation of the said ConnectionState
+func (s ConnectionState) String() string {
+	switch s {
+	case idle:
+		return "IDLE"
+	case connecting:
+		return "CONNECTING"
+	case ready:
+		return "READY"
+	case shutdown:
+		return "SHUTDOWN"
+	default:
+		return "Invalid-State"
+	}
+}
+
+const (
+	// idle indicates the conn is idle.
+	idle ConnectionState = iota
+
+	// connecting indicates the conn is connecting.
+	connecting
+
+	// disconnected indicates the conn marked to be picked by reconnect handler.
+	disconnected
+
+	// ready indicates the conn is ready for work.
+	ready
+
+	// shutdown indicates the ClientConn has started shutting down.
+	shutdown
+)
+
+// HostAddress represents a server the client connects to.
+type HostAddress struct {
+	// Host is the IP address and port of the RPC server you want to connect
+	// to.
+	Host string
+
+	// Endpoint is the websocket endpoint on the RPC server.  This is
+	// typically "ws".
+	Endpoint string
+
+	// retryCount holds the number of times the client has tried to
+	// reconnect to the RPC server.
+	retryCount int64
+}
+
+// Balancer uses Picker to pick connections.
+// It collects and maintains the connectivity states.
+type Balancer interface {
+	// NextConn gets the next usabel connection.
+	NextConn() (wsConn *websocket.Conn, hostAddress *HostAddress, err error)
+
+	// ConnectionState gets the connectionstate of corresponding wsConn.
+	ConnectionState(host string) (state ConnectionState, ok bool)
+
+	// WsConnection gets the wsConn corresponding to the host.
+	WsConnection(host string) (wsConn *websocket.Conn, connOk bool)
+
+	// NextDisconnectedWsConn will iterate over the connection state map and
+	// return the first wsConnection that has its state as Disconnected.
+	NextDisconnectedWsConn() *HostAddress
+
+	// NotifyConnStateChange is called by rpcClient when the ConnectionState
+	// for a hostaddress changes.
+	NotifyConnStateChange(sc *HostAddress, state ConnectionState, sync bool)
+
+	// UpdateReconnectAttempt will increase retryattempt + 1.
+	UpdateReconnectAttempt(hostAdd *HostAddress)
+
+	// NotifyReconnect will update the map for wsConns and update the connection state
+	// for this address.
+	NotifyReconnect(wsConn *websocket.Conn, hostAdd *HostAddress)
+
+	// Close closes the passed ws connection and sets the state as Disconnected
+	// by calling NotifyConnStateChange.
+	// Returns true if this was the last WS connection to be closed.
+	Close(wsConn *websocket.Conn) bool
+
+	// CloseAll closes all the ws connections which are active at present.
+	CloseAll()
+}
+
+// Picker is used to pick the next connection to be used.
+type Picker interface {
```

Comment:
This should not be exported.  It is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.